### ISSUE TITLE
Loss computation

### ISSUE THREAD

**Issue by yd-yin** [PrimitiveAnything/PrimitiveAnything#9]
Hi, thanks for your impressive and insightful work!

I have a few questions about the implementation of the losses:

1 Gumbel-Softmax

1.1 Am I understanding correctly? The model predicts parameters with dependency (c -> t -> r -> s), e.g., t = decoder(embed, e(c)). To allow gradients to flow through c, Gumbel-Softmax is applied to convert c_logit into c_gs (a vector, instead of c_argmax, a scalar). Then e(c_gs) is computed as something like c_gs @ embedding_matrix, which participates in the decoding process.

1.2 What's the strategy of the temperature in Gumbel-softmax (fixed, annealed, etc.)?

2 n samples

2.1 The paper mentions that n samples are generated. What is the purpose of generating multiple samples?

2.2 How are the losses from the n samples combined? For example, e.g., minimum-loss strategy, or something else?

Thanks a lot!

**Comment by hyz317** [comment]
Hi,

Thank you for your interest in our work!

- **Gumbel-Softmax**
  During training, we use the ground-truth attribute for the current primitive to look up the corresponding embedding directly from the learned embeddings. During inference, we directly use the sampled attribute to look up its embedding. The temperature for Gumbel-Softmax is fixed at 0.2.

- **'n' samples**
  This appears to be a slight misunderstanding of the terminology in the paper. When we refer to 'n', we are not talking about generating *n* different assembly samples. Instead, 'n' refers to the number of primitives that constitute a single "primitive assembly." For an assembly with *n* primitives (p1, p2, ..., pn), the transformer's input is a sequence like [SOS], p1, p2, ..., p(n-1), and it's trained to output the shifted sequence p1, p2, ..., pn, [EOS]. We then collect all the next-token outputs to form an assembly, sample point clouds and calculate Chamfer Distance loss at the assembly level.

**Comment by yd-yin** [comment]
Yeah, thanks for the helpful response!
